engines/lab/resource.cpp

```diff
@@ -0,0 +1,344 @@
+		_vm->_rooms[i]._transitionType = dataFile->readByte();
+
+		_vm->_rooms[i]._view[NORTH] = nullptr;
+		_vm->_rooms[i]._view[SOUTH] = nullptr;
+		_vm->_rooms[i]._view[EAST] = nullptr;
+		_vm->_rooms[i]._view[WEST] = nullptr;
+		_vm->_rooms[i]._rules = nullptr;
+		_vm->_rooms[i]._roomMsg = nullptr;
+	}
+
+	delete dataFile;
+	return true;
+}
+
+InventoryData *Resource::readInventory(const char *fileName) {
+	Common::File *dataFile = openDataFile(fileName, MKTAG('I', 'N', 'V', '1'));
+
+	_vm->_numInv = dataFile->readUint16LE();
+	InventoryData *inventory = new InventoryData[_vm->_numInv + 1];
+
+	for (uint16 i = 1; i <= _vm->_numInv; i++) {
+		inventory[i]._many = dataFile->readUint16LE();
+		inventory[i]._name = readString(dataFile);
+		inventory[i]._bitmapName = readString(dataFile);
+	}
+
+	delete dataFile;
+
+	return inventory;
+}
+
+bool Resource::readViews(uint16 roomNum) {
+	Common::String fileName = "LAB:Rooms/" + Common::String::format("%d", roomNum);
+	Common::File *dataFile = openDataFile(fileName.c_str(), MKTAG('R', 'O', 'M', '4'));
+
+	_vm->_rooms[roomNum]._roomMsg = readString(dataFile);
+	_vm->_rooms[roomNum]._view[NORTH] = readView(dataFile);
+	_vm->_rooms[roomNum]._view[SOUTH] = readView(dataFile);
+	_vm->_rooms[roomNum]._view[EAST] = readView(dataFile);
+	_vm->_rooms[roomNum]._view[WEST] = readView(dataFile);
+	_vm->_rooms[roomNum]._rules = readRule(dataFile);
+
+	_vm->_music->updateMusic();
+
+	delete dataFile;
+	return true;
+}
+
+Common::String Resource::translateFileName(Common::String filename) {
+	filename.toUppercase();
+	Common::String fileNameStrFinal;
+
+	if (filename.hasPrefix("P:")) {
+		if (_vm->_isHiRes)
+			fileNameStrFinal = "GAME/SPICT/";
+		else
+			fileNameStrFinal = "GAME/PICT/";
+	} else if (filename.hasPrefix("LAB:"))
+		fileNameStrFinal = "GAME/";
+	else if (filename.hasPrefix("MUSIC:"))
+		fileNameStrFinal = "GAME/MUSIC/";
+
+	if (filename.contains(':')) {
+		while (filename[0] != ':') {
+			filename.deleteChar(0);
+		}
+
+		filename.deleteChar(0);
+	}
+
+	fileNameStrFinal += filename;
+
+	return fileNameStrFinal;
+}
+
+Common::File *Resource::openDataFile(const char *fileName, uint32 fileHeader) {
+	Common::File *dataFile = new Common::File();
+	dataFile->open(translateFileName(fileName));
+	if (!dataFile->isOpen())
+		error("openDataFile: Couldn't open %s (%s)", translateFileName(fileName).c_str(), fileName);
+
+	if (fileHeader > 0) {
+		uint32 headerTag = dataFile->readUint32BE();
+		if (headerTag != fileHeader) {
+			dataFile->close();
+			error("openDataFile: Unexpected header in %s (%s) - expected: %d, got: %d", translateFileName(fileName).c_str(), fileName, fileHeader, headerTag);
+		}
+	}
+
+	return dataFile;
+}
+
+char *Resource::readString(Common::File *file) {
+	byte size = file->readByte();
+	if (!size)
+		return NULL;
+	char *str = new char[size];
+	char *c = str;
+	for (int i = 0; i < size; i++) {
+		*c = file->readByte();
+		// Decrypt char
+		*c++ = (i < size - 1) ? *c - 95 : '\0';
```

Comment:
Actually, this seems to be why the game crashes for me, right at the end of the intro. If I change it to

*c = (i < size - 1) ? *c - 95 : '\0';
c++;

It works for me. Without that, I get some message about opening a file where the name is all garbage characters.